pkg/controller/directvolumemigration/resource.go

```diff
@@ -0,0 +1,352 @@
+		finalLimitRange.Spec.Limits = append(finalLimitRange.Spec.Limits, *containerLimitRangeItem)
+	}
+	return finalLimitRange, nil
+}
+
+// mergeResourceLimitRangeItems given target and source LimitRangeItems, merges source into target.
+// it compares the Max & Min resource limits of CPU and memory and keeps the most restrictive value
+// NOTE: does not consider MaxLimitRequestRatio for comparison
+func mergeResourceLimitRangeItems(target *corev1.LimitRangeItem, source *corev1.LimitRangeItem) {
+	if target != nil && source != nil {
+		if source.Max != nil {
+			// for max limits, the most restrictive value is the one which is the minimum
+			if target.Max.Cpu().Cmp(*source.Max.Cpu()) > 0 {
+				target.Max[corev1.ResourceCPU] = source.Max[corev1.ResourceCPU]
+			}
+			if target.Max.Memory().Cmp(*source.Max.Memory()) > 0 {
+				target.Max[corev1.ResourceMemory] = source.Max[corev1.ResourceMemory]
+			}
+		}
+		if source.Min != nil {
+			// for min limits, the most restrictive value is the one which is the maximum
+			if target.Min.Cpu().Cmp(*source.Min.Cpu()) < 0 {
+				target.Min[corev1.ResourceCPU] = source.Min[corev1.ResourceCPU]
+			}
+			if target.Min.Memory().Cmp(*source.Min.Memory()) < 0 {
+				target.Min[corev1.ResourceMemory] = source.Min[corev1.ResourceMemory]
+			}
+		}
+	}
+}
+
+// applyLimitRangeItemOnRequirements given a ResourceRequirement and a LimitRangeItem, applies limits on the given ResourceRequirement
+// the ResourceRequirement is modified in-place such that no resource value violates the limits in the given LimitRangeItem
+func applyLimitRangeItemOnRequirements(requirements *corev1.ResourceRequirements, limit corev1.LimitRangeItem) {
+	if requirements.Requests != nil {
+		if limit.Max != nil {
+			//  apply Max limits on Requests and Limits, none should exceed the Max
+			if requirements.Requests.Cpu().Cmp(*limit.Max.Cpu()) > 0 {
+				requirements.Requests[corev1.ResourceCPU] = *limit.Max.Cpu()
+			}
+			if requirements.Requests.Memory().Cmp(*limit.Max.Memory()) > 0 {
+				requirements.Requests[corev1.ResourceMemory] = *limit.Max.Memory()
+			}
+		}
+		if limit.Min != nil {
+			// apply Min limits on Requests and Limits, none should be less than the Min
+			if requirements.Requests.Cpu().Cmp(*limit.Min.Cpu()) < 0 {
+				requirements.Requests[corev1.ResourceCPU] = *limit.Min.Cpu()
+			}
+			if requirements.Requests.Memory().Cmp(*limit.Min.Memory()) < 0 {
+				requirements.Requests[corev1.ResourceMemory] = *limit.Min.Memory()
+			}
+		}
+	}
+	if requirements.Limits != nil {
+		if limit.Max != nil {
+			//  apply Max limits on Requests and Limits, none should exceed the Max
+			if requirements.Limits.Cpu().Cmp(*limit.Max.Cpu()) > 0 {
+				requirements.Limits[corev1.ResourceCPU] = *limit.Max.Cpu()
+			}
+			if requirements.Limits.Memory().Cmp(*limit.Max.Memory()) > 0 {
+				requirements.Limits[corev1.ResourceMemory] = *limit.Max.Memory()
+			}
+		}
+	}
+	// if limits are smaller than requests, delete limit values
```

Comment:
Could this result in a pod failing to start? I know there are issues when resources aren't specified but there's a resource quota in place. I would be concerned that in the case where there is a ResourceQuota with cpu/memory limits set the above may result in failed pods. Backing up a step here, the only way limits could be smaller than requests here would be if the limit ranges are doing something bizarre, though right? i.e. limit ranges with limit values lower than request values? Should we error out here, or should we just do what you're doing -- and in the edge case where irrational limit ranges intersect with resource quotas, we'll just fail to start pods?

Reply:
@sseago Yes, if `resources.Limits` > `resources.Requests`, Pods fail to start. Note that this only happens when there are `LimitRange` present in a namespace. We are taking defensive approach here. Basically, by deleting Limits altogether, we are letting the `LimitRange` decide what the maximum should be. As long as `resources.Requests` remains under limits, the `resources.Limits` can be safely deleted such that it can be managed automatically by Kube. So this code exists such that we never fail creating Pods if Limits are insane. Also, I think this will be a corner case. 

When `LimitRange` are not present, we still fallback on the User configured values then the defaults in that order.